src/main/java/frc/robot/subsystems/swerve/SwerveModule.java

```diff
@@ -0,0 +1,264 @@
+            upperBound = scopeReference + (360 - lowerOffset);
+        } else {
+            upperBound = scopeReference - lowerOffset;
+            lowerBound = scopeReference - (360 + lowerOffset);
+        }
+        while(newAngle < lowerBound) {
+            newAngle += 360;
+        }
+        while(newAngle > upperBound) {
+            newAngle -= 360;
+        }
+        if(newAngle - scopeReference > 180) {
+            newAngle -= 360;
+        } else if(newAngle - scopeReference < -180) {
+            newAngle += 360;
+        }
+        return newAngle;
+    }
+
+    public void setDesiredState(SwerveModuleState desiredState, boolean isOpenLoop) {
+        // Custom optimize command, since default WPILib optimize assumes
+        // continuous controller which CTRE is not
+        desiredState = optimize(
+                desiredState,
+                getState().angle
+        );
+        if(isOpenLoop) {
+            /* If we're in open loop, we don't want to use the PID controller,
+            and we just set the drive speed in percentage of the max speed */
+            double percentOutput =
+                    desiredState.speedMetersPerSecond / SwerveConstants.MAX_SPEED;
+            mDriveMotor.set(
+                    ControlMode.PercentOutput,
+                    percentOutput / SwerveConstants.SPEED_DIVIDER);
+        } else {
+            /* If we're in closed loop, we want to use the PID controller,
+            and we set the drive speed in meters per second  */
+            double velocity = EncoderConversions.MPSToFalcon(
+                    desiredState.speedMetersPerSecond,
+                    SwerveConstants.WHEEL_CIRCUMFERENCE,
+                    SwerveConstants.DRIVE_GEAR_RATIO);
+            // Sets the drive motor velocity with the driveFeedforward.
+            mDriveMotor.set(
+                    ControlMode.Velocity, velocity, DemandType.ArbitraryFeedForward,
+                    driveFeedforward.calculate(desiredState.speedMetersPerSecond));
+        }
+
+        /*//Prevent rotating module if speed is less than 1%. Prevents jitter.
+        double desiredAngle =
+                (Math.abs(desiredState.speedMetersPerSecond) <= (SwerveConstants
+                .MAX_SPEED * 0.01)) ?
+                lastAngle :
+                desiredState.angle.getDegrees(); */
+        double desiredAngle = desiredState.angle.getDegrees();
+        // Set the angle motor's position to the desired angle.
+        mAngleMotor.set(
+                ControlMode.Position, EncoderConversions.degreesToFalcon(
+                        desiredAngle,
+                        SwerveConstants.ANGLE_GEAR_RATIO));
+        lastDesiredState = desiredState;
+    }
+
+    /**
+     * Sets the value of the integrated angle encoder to the absolute value of the
+     * angle,
+     * from the absolute encoder, and the absolute encoder offset.
+     */
+    private void resetToAbsolute() {
```

Comment:
what about using the mag? have you looked into it yet?

Reply:
haven't